src/main/java/org/truffleruby/language/methods/InternalMethod.java

```diff
@@ -233,7 +232,6 @@ public RubyProc getProc() {
 
     public RootCallTarget getCallTarget() {
         if (callTarget == null) {
-            CompilerDirectives.transferToInterpreterAndInvalidate();
```

Comment:
This it not OK, it would kill performance (e.g. the guard in CallInternalMethodNode), the transfer needs to remain here.

I think a `CompilerDirectives.transferToInterpreterAndInvalidate();` in CachedSupplier#get just before the `synchronized` makes sense.

Reply:
> it would kill performance (e.g. the guard in CallInternalMethodNode)

It's just a later transfer, and only when needed. Can you show how that kills performance?

> in `CachedSupplier#get`

But I don't think that's right for all users of `in CachedSupplier#get` - for example it isn't needed for the non-lazy-translation case.

Reply:
```java
public abstract class CallInternalMethodNode extends RubyBaseNode {
//...
    @Specialization(
            guards = {
                    "isSingleContext()",
                    "method.getCallTarget() == cachedCallTarget",
                    "!cachedMethod.alwaysInlined()" },
            assumptions = "getMethodAssumption(cachedMethod)", // to remove the inline cache entry when the method is redefined or removed
            limit = "getCacheLimit()")
    protected Object callCached(
```
That guard calls InternalMethod#getCallTarget(), anything more expensive than the existing single read (+ maybe null guard, but no merge) makes it slower.
In this case you are potentially doing an interface call to `Supplier#get` (inside `CachedSupplier#get`), i.e. a non-inlined call.

Note that in single-context, the `method` (an InternalMethod) might be PE-constant (but not always, only if monomorphic), but in multi-contexts the `method` is never PE-constant and the condition would never fold and the call would always be in the graph (you can check multi-contexts with `--engine.ForceCodeSharing --experimental-engine-caching`).

CachedSupplier shouldn't `implements Supplier<T>` that's just confusing and makes it harder to find usages, could you remove that in this PR for clarity?

> But I don't think that's right for all users of in CachedSupplier#get - for example it isn't needed for the non-lazy-translation case.

It's not strictly needed, but it also doesn't make sense to compile a generic call to `Supplier#get` on a fast path.
If we run into this transferToInterpreterAndInvalidate(), we have a bug, specifically that the `callTarget` is not set as expected (the first commit fixes such a case).

Reply:
> could you remove that in this PR for clarity?

Done.

Reply:
I think you are hoping that SVM can see the 3 possible `Supplier` which are stored in a CachedSupplier, and to automatically virtual-inline the 3 implementations. There is no such thing, SVM doesn't track this and it could be any Supplier in the image, and PE wouldn't/can't virtual-inline like this interface calls.

Reply:
> I think you are hoping that SVM can see the 3 possible Supplier which are stored in a CachedSupplier, and to automatically virtual-inline the 3 implementations.

All uses of suppliers are slow-path though.

Reply:
What does that impact? Just the code volume? They aren't run unless we're in the slow path are they?

Reply:
Well it's obviously slower, for all cases where the InternalMethod is not PE-constant, which is not only slow paths as said above.

Reply:
BTW, CachedLazyCallTargetSupplier#value is a volatile read, we definitely don't want any of that in the compilation graph when we can avoid it.

Reply:
Right. This weird case wouldn't be solved by any of the 2 approaches:
```ruby
def foo(o)
  o.baz
end

class Klass
  def bar
    14
  end
end

objects = 1000.times.map {
  o = Klass.new
  class << o
    alias_method :baz, :bar
  end
  o
}

loop do
  objects.each { |o| foo(o) }
end
```
I think the only way for that one would be to have an extra read (potentially behind a `if`) all the time, and I don't think that's worth it. (this is a case of semantically a megamorphic lookup/call, it should be very rare in practice)
For such a program, one could always disable lazy translation (which is what multi-contexts does, for other reasons), or we could eagerly resolve CallTargets on `alias/public/...` since we can only propagate to the CachedLazyCallTargetSupplier, not to all InternalMethods holding that CachedLazyCallTargetSupplier instance.

Reply:
It doesn't, I was too kind with 1000 but with 1000000 I can see many deopts on this branch.
Once the first `loop` iteration is done it's fine of course, so it is not an infinite deopt, just many of them.

Reply:
I'm losing track of what you want from me again.

We want (presumably) to fix these three programs so they don't run in de-opt loops:

```ruby
def foo(o)
  o.bar
end

loop do
  o = Object.new
  def o.bar
    14
  end
  foo(o)
end
```

```ruby
def foo(o)
  o.baz
end

class Klass
  def bar
    14
  end
end

loop do
  o = Klass.new
  class << o
    alias_method :baz, :bar
  end
  foo(o)
end
```

```ruby
def foo(o)
  o.baz
end

class Klass
  def bar
    14
  end
end

objects = 1000.times.map {
  o = Klass.new
  class << o
    alias_method :baz, :bar
  end
  o
}

loop do
  objects.each { |o| foo(o) }
end
```

On 56458ee61c66459625a6e301cd1231e7e1d818d9, all three programs stabilise.

With your patch:

```diff
diff --git a/src/main/java/org/truffleruby/language/methods/InternalMethod.java b/src/main/java/org/truffleruby/language/methods/InternalMethod.java
index ebd1066c588..04f36622d38 100644
--- a/src/main/java/org/truffleruby/language/methods/InternalMethod.java
+++ b/src/main/java/org/truffleruby/language/methods/InternalMethod.java
@@ -170,12 +170,6 @@ public class InternalMethod implements ObjectGraphNode {
         this.proc = proc;
         this.callTarget = callTarget;
         this.callTargetSupplier = callTargetSupplier;
-
-        /* If the call target supplier has already been run, then don't wait until the first time the InternalMethod is
-         * asked for the call target, because this would be a deoptimization in getCallTarget(). */
-        if (callTarget == null && callTargetSupplier != null) {
-            this.callTarget = callTargetSupplier.getWhenAvailable();
-        }
     }
 
     public SharedMethodInfo getSharedMethodInfo() {
diff --git a/src/main/java/org/truffleruby/language/methods/LiteralMethodDefinitionNode.java b/src/main/java/org/truffleruby/language/methods/LiteralMethodDefinitionNode.java
index 6e624e183a2..3b37448a936 100644
--- a/src/main/java/org/truffleruby/language/methods/LiteralMethodDefinitionNode.java
+++ b/src/main/java/org/truffleruby/language/methods/LiteralMethodDefinitionNode.java
@@ -78,7 +78,7 @@ public class LiteralMethodDefinitionNode extends RubyContextSourceNode {
                 false,
                 null,
                 null,
-                null,
+                callTargetSupplier.getWhenAvailable(),
                 callTargetSupplier);
 
         if (isDefSingleton) {
```

Programs 1 and 3 are ok, but 2 runs in a de-opt loop. That's not what we wanted. I don't understand 'This weird case wouldn't be solved' becuase it seems fine to me already with both of our proposed solutions.

Can we refocus on what are we trying to fix now? What does your patch achieve?

Reply:
I've already approved the PR and it's already in the merge queue, there is nothing left to do.
The 3rd one stabilizes but after many deopts with a large enough N for `N.times.map`, IMHO it's fine as-is, I just wanted to mention neither approach helps for that case, because all definitions happen before the usages, hence the check in InternalMethod's constructor does not help there (until we reach the second iteration of the `loop`, then it does help, probably I should have removed that loop to clarify).